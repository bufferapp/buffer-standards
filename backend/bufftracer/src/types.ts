export const TAG_NAME = 'api'
export const TAG_LIMIT = 200 // This is the character limit including all the characters of the string `key:value`
export const TAG_KEY_LENGTH = `${TAG_NAME}.fields10:`.length // This is the string length of the `key:` part
export const TAG_VALUE_LIMIT = TAG_LIMIT - TAG_KEY_LENGTH // This is the remaining character lenth we can use to set fields value
export const TAG_SEPARATOR = '_' // DD will join array of string using "_" as a separator

// We need several fields properties because DD APM tags are limited to 200 characters including the key and value length:
// This results in tag being cut and incomplete data for us because the tag won't contain all fields being queried
// api.fields:account.id_account.currentorganization_currentorganization.id_currentorganization.isonebufferorganization_currentorganization.name_currentorganization.role_currentorganization.billing_billi
export type APIMetadataFields = {
  fields?: Array<string>
  fields2?: Array<string>
  fields3?: Array<string>
  fields4?: Array<string>
  fields5?: Array<string>
  fields6?: Array<string>
  fields7?: Array<string>
  fields8?: Array<string>
  fields9?: Array<string>
  fields10?: Array<string>
}

export type APIRequestMetadata = {
  name: 'core-api' | string
  type: 'graphql' | 'rpc' | 'rest'
  client: 'buffertools-graphql-playground' | string
  edge: string
  deprecatedFields: Array<string>
  args: Array<string>
} & APIMetadataFields

type Request<P, B> = {
  headers: Record<string, string>
  params: P
  body: B
}
export type GraphQlRequest = Request<Record<string, never>, { query: string }>
export type RPCRequest = Request<
  { method?: string },
  { name?: string; args?: string | Record<string, unknown> }
>
