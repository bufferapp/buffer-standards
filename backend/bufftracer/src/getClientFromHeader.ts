import { GraphQlRequest, RPCRequest } from './types'

export const HEADER_CLIENT_ID = 'x-buffer-client-id'
export const HEADER_CLIENT_UNKNOWN = 'unknown'

export function getClientFromHeader(args: {
  req: RPCRequest | GraphQlRequest
}): string {
  const { req } = args

  return req.headers[HEADER_CLIENT_ID] || HEADER_CLIENT_UNKNOWN
}
