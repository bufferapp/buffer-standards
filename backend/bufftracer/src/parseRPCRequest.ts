import { APIRequestMetadata, RPCRequest } from './types'
import { getClientFromHeader } from './getClientFromHeader'
import * as env from 'env-var'

export function parseRPCRequest(args: { req: RPCRequest }): APIRequestMetadata {
  const { req } = args

  const rpcMethod = req.body.name || req.params.method
  const edge = `rpc ${rpcMethod}`

  const rpcArgs =
    typeof req.body.args === 'string'
      ? JSON.parse(req.body.args)
      : req.body.args || {}

  const argumentList = Object.keys(rpcArgs).map(
    (argumentName) => `rpc.${rpcMethod}.${argumentName}`,
  )

  return {
    name: env.get('DD_SERVICE_NAME').required().asString(),
    type: 'rpc',
    client: getClientFromHeader({ req }),
    edge,
    fields: [],
    deprecatedFields: [],
    args: argumentList,
  }
}
