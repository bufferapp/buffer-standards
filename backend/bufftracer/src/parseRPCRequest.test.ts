import { RPCRequest } from './types'
import { parseRPCRequest } from './parseRPCRequest'
import { HEADER_CLIENT_ID } from './getClientFromHeader'

describe('parseRPCRequest', () => {
  it('returns metadata from request with name in body', () => {
    // Arrange
    const req: RPCRequest = {
      headers: {
        [HEADER_CLIENT_ID]: 'webapp-account',
      },
      params: {},
      body: {
        name: 'getAccount',
        args: JSON.stringify({
          email: 'test@buffer.com',
        }),
      },
    }

    // Act
    const result = parseRPCRequest({ req })

    // Assert
    expect(result).not.toBeNull()
    expect(result).toMatchSnapshot()
  })

  it('returns metadata from request with method in params', () => {
    // Arrange
    const req: RPCRequest = {
      headers: {
        [HEADER_CLIENT_ID]: 'webapp-account',
      },
      params: {
        method: 'getAccount',
      },
      body: {
        args: JSON.stringify({
          email: 'test@buffer.com',
        }),
      },
    }

    // Act
    const result = parseRPCRequest({ req })

    // Assert
    expect(result).not.toBeNull()
    expect(result).toMatchSnapshot()
  })

  it('returns metadata from request with args as null', () => {
    // Arrange
    const req: RPCRequest = {
      headers: {
        [HEADER_CLIENT_ID]: 'webapp-account',
      },
      params: {
        method: 'getAccount',
      },
      body: {},
    }

    // Act
    const result = parseRPCRequest({ req })

    // Assert
    expect(result).not.toBeNull()
    expect(result).toMatchSnapshot()
  })

  it('returns metadata from request with args as object', () => {
    // Arrange
    const req: RPCRequest = {
      headers: {
        [HEADER_CLIENT_ID]: 'webapp-account',
      },
      params: {
        method: 'getAccount',
      },
      body: {
        args: {
          email: 'test@buffer.com',
        },
      },
    }

    // Act
    const result = parseRPCRequest({ req })

    // Assert
    expect(result).not.toBeNull()
    expect(result).toMatchSnapshot()
  })
})
