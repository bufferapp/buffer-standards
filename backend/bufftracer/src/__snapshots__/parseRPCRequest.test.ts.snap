// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseRPCRequest returns metadata from request with args as null 1`] = `
Object {
  "args": Array [],
  "client": "webapp-account",
  "deprecatedFields": Array [],
  "edge": "rpc getAccount",
  "fields": Array [],
  "name": "foo-api",
  "type": "rpc",
}
`;

exports[`parseRPCRequest returns metadata from request with args as object 1`] = `
Object {
  "args": Array [
    "rpc.getAccount.email",
  ],
  "client": "webapp-account",
  "deprecatedFields": Array [],
  "edge": "rpc getAccount",
  "fields": Array [],
  "name": "foo-api",
  "type": "rpc",
}
`;

exports[`parseRPCRequest returns metadata from request with method in params 1`] = `
Object {
  "args": Array [
    "rpc.getAccount.email",
  ],
  "client": "webapp-account",
  "deprecatedFields": Array [],
  "edge": "rpc getAccount",
  "fields": Array [],
  "name": "foo-api",
  "type": "rpc",
}
`;

exports[`parseRPCRequest returns metadata from request with name in body 1`] = `
Object {
  "args": Array [
    "rpc.getAccount.email",
  ],
  "client": "webapp-account",
  "deprecatedFields": Array [],
  "edge": "rpc getAccount",
  "fields": Array [],
  "name": "foo-api",
  "type": "rpc",
}
`;
